File: js/src/lib/journeys/journeyRunner.ts
import { expect, Locator, Page } from '@playwright/test';
import { TestUser } from '../fixtures/testUsers.js';
import { isJourneyId, JourneyId, journeySpecs } from './generatedJourneySpecs.js';
import type { FixtureValue, SelectorSpec, TextPattern } from './journeyTypes.js';

const toRegExp = ({ pattern, ignoreCase }: TextPattern): RegExp => new RegExp(pattern, ignoreCase ? 'i' : '');

const resolveLocator = (page: Page, selector: SelectorSpec): Locator => {
  let locator: Locator;

  switch (selector.by) {
    case 'role': {
      const options = {
        ...(selector.name === null ? {} : { name: toRegExp(selector.name) }),
        ...(selector.level === null ? {} : { level: selector.level })
      };
      locator = page.getByRole(selector.role, options);
      break;
    }
    case 'label':
      locator = page.getByLabel(toRegExp(selector.name));
      break;
    case 'testId':
      locator = page.getByTestId(selector.value);
      break;
  }

  return selector.first ? locator.first() : locator;
};

const resolveFixtureValue = (user: TestUser, value: FixtureValue): string => {
  switch (value) {
    case 'user.email':
      return user.email;
    case 'user.password':
      return user.password;
    case 'user.displayName':
      if (user.displayName === undefined) {
        throw new Error('Journey requires user.displayName, but the fixture value is undefined');
      }
      return user.displayName;
  }
};

export const executeJourney = async (
  page: Page,
  user: TestUser,
  journeyId: JourneyId,
  activeJourneyIds: readonly JourneyId[] = []
): Promise<void> => {
  if (activeJourneyIds.includes(journeyId)) {
    throw new Error(`Journey inheritance cycle while executing ${journeyId}`);
  }

  const spec = journeySpecs[journeyId];
  if (spec.extends !== null) {
    if (!isJourneyId(spec.extends)) {
      throw new Error(`Journey ${journeyId} extends unknown journey: ${spec.extends}`);
    }
    await executeJourney(page, user, spec.extends, [...activeJourneyIds, journeyId]);
  }

  for (const step of spec.steps) {
    switch (step.action) {
      case 'navigate':
        await page.goto(step.path);
        break;
      case 'click':
        await resolveLocator(page, step.selector).click();
        break;
      case 'fill':
        await resolveLocator(page, step.selector).fill(resolveFixtureValue(user, step.value));
        break;
      case 'assert-visible':
        await expect(resolveLocator(page, step.selector)).toBeVisible();
        break;
    }
  }
};

File: py/src/playwright_adventures/journeys/journey_runner.py
import re

from playwright.async_api import Locator, Page, expect

from ..config import get_base_url
from .generated_specs import JOURNEY_SPECS, JourneyId, is_journey_id
from .journey_types import FixtureValue, SelectorSpec, TextPattern
from .models import TestUser


def _text_pattern(value: TextPattern) -> re.Pattern[str]:
    flags = re.IGNORECASE if value["ignoreCase"] else 0
    return re.compile(value["pattern"], flags)


def _locator(page: Page, selector: SelectorSpec) -> Locator:
    if selector["by"] == "role":
        name = None if selector["name"] is None else _text_pattern(selector["name"])
        if selector["level"] is None:
            locator = page.get_by_role(selector["role"], name=name)
        else:
            locator = page.get_by_role(selector["role"], name=name, level=selector["level"])
    elif selector["by"] == "label":
        locator = page.get_by_label(_text_pattern(selector["name"]))
    else:
        locator = page.get_by_test_id(selector["value"])

    return locator.first if selector["first"] else locator


def _fixture_value(user: TestUser, value: FixtureValue) -> str:
    if value == "user.email":
        return user.email
    if value == "user.password":
        return user.password
    if user.display_name is None:
        raise ValueError("Journey requires user.displayName, but the fixture value is undefined")
    return user.display_name


async def execute_journey(
    page: Page,
    user: TestUser,
    journey_id: JourneyId,
    active_journey_ids: tuple[JourneyId, ...] = (),
) -> None:
    if journey_id in active_journey_ids:
        raise ValueError(f"Journey inheritance cycle while executing {journey_id}")

    spec = JOURNEY_SPECS[journey_id]
    parent_id = spec["extends"]
    if parent_id is not None:
        if not is_journey_id(parent_id):
            raise ValueError(f"Journey {journey_id} extends unknown journey: {parent_id}")
        await execute_journey(page, user, parent_id, (*active_journey_ids, journey_id))

    for step in spec["steps"]:
        if step["action"] == "navigate":
            await page.goto(f"{get_base_url()}{step['path']}")
        elif step["action"] == "click":
            await _locator(page, step["selector"]).click()
        elif step["action"] == "fill":
            await _locator(page, step["selector"]).fill(_fixture_value(user, step["value"]))
        else:
            await expect(_locator(page, step["selector"])).to_be_visible()
